### ISSUE TITLE
Torque OSWEC hydraulic PTO

### ISSUE THREAD

**Issue by NicolasQu** [WEC-Sim/WEC-Sim#247]
Hi!
I have a question about the hydraulic PTO for the OSWEC example.
When I increase the values of the piston top and bottom area, I saw that the power keeps increasing.
When I plot the velocity of the flap in pitch together with the torque applied by the PTO (calculated inside 'rotary to linear adjustable rod'), I noticed that the **torque is positive when the velocity is positive**, and **negative when the pitch velocity is negative**. Shouldn't this be the other way around?
Thanks in advance!
Regards
Nicolas

**Comment by kmruehl** [comment]
@NicolasQu, I looked into this issue a bit, and I think this has to do with a few different factors. 

The first being that all outputs from joints/constraints are reported in the [body fixed local coordinate system](http://wec-sim.github.io/WEC-Sim/code_structure.html#ptos), not the global coordinate system. If you go under the mask for the PTO blocks you will see a series of Simscape Multibody [Rigid Transform](https://www.mathworks.com/help/physmod/sm/ref/rigidtransform.html) blocks in combination with a Simscape Multibody [Revolute Joint ](https://www.mathworks.com/help/physmod/sm/ref/revolutejoint.html) (for the OSWEC) and a Simscape Multibody [Prismatic Joint ](https://www.mathworks.com/help/physmod/sm/ref/prismaticjoint.html) (for the RM3). Simscape Multibody specifies a default DOF for each joint, and WEC-Sim rotates these joints to the appropriate orientation. As a result, depending on where you are looking at the signal, it may not have been re-oriented yet. However, all dynamics/loads saved to the output data structure should have the correct body-fixed orientation. 

Based on your post, the most likely reason for the observed sign discrepancy is that the `output.ptos.forceActuation` signal is the signal applied to the pto, not the signal used to calculate power. Power is calculated in the `OSWEC_Hydraulic_PTO/Rotational PTO Actuation Torque/InternalMechanics` subsystem (for the OSWEC) by:
 `output.ptos.powerInternalMechanics`=`output.bodies.velocity`^2 * `pto.c`

WEC-Sim actually does not directly output the internal force due to damping alone because:
`output.ptos.forceInternalMechanics` = `output.bodies.position` * `pto.k` + `output.bodies.velocity` * `pto.c` 

If your model has `pto.k = 0` then you can use `output.ptos.forceInternalMechanics` to reconstruct `output.ptos.powerInternalMechanics`.

**Comment by NicolasQu** [comment]
@kmruehl , thank you for your explanation.

There is however still something I don't understand.
If I calculate the absorbed power for the OSWEC device with hydraulic PTO, the power seems to rise when the piston area rises. Calculations are done for regular waves with H = 1.5 m and T = 7s. In the figure below, the power reaches values until 600 kW which seems very high to me. I calculated the power by taking the mean value of output.ptosim.pistonNCF.absPower/1e3. pto.c and pto.k are set to 0.
![image](https://user-images.githubusercontent.com/38635488/40969577-8d80f690-68b8-11e8-8a92-172861b63217.png)


When I do the same for the RM3 device with hydraulic PTO, I found an optimal value for the piston area with an optimal mean absorbed power of 105 kW, which seems more logical.
![image](https://user-images.githubusercontent.com/38635488/40969609-a1725c34-68b8-11e8-8c55-2fd2b965190f.png)

**Comment by kmruehl** [comment]
@NicolasQu it does seem unusual that the power for the OSWEC is increasing and does not show an optimal value based on piston area (within the plotted range). These application cases were developed to demonstrate functionality of the WEC-Sim code, but the devices and ptos were not optimized so it's hard to provide insight. I recommend running a larger range of piston areas for the OSWEC because I would expect its loads to be larger than those on the RM3. The other important factor to note is that the OSWEC model doesn't account for friction or drag which will likely be dominating physics, and thus WEC-Sim will overpredict the power output with their omission.

**Comment by kmruehl** [comment]
@NicolasQu does this answer your question?

**Comment by NicolasQu** [comment]
@kmruehl the loads are indeed larger for the OSWEC than for the RM3, but the absorbed power going to 600 kW seems very high, especially when comparing with the absorbed power for OSWEC with a linear damper which goes to ca 260 kW. If the range of piston areas increased, I get an error for excessive pitch angles.

When I added one minus sign in the Simulink model calculation of the PTO-torque of the hydraulic OSWEC, this is the result I got:
![image](https://user-images.githubusercontent.com/38635488/41200479-553f2eaa-6ca5-11e8-86ec-b1a274054bf5.png)
In this case the power reaches a maximum of ca 225 kW, which seems more logical. Due to adding this minus sign, the forceActuation(:,5) and velocity(:,5) now always have opposite signs.

**Comment by kmruehl** [comment]
@NicolasQu, sorry for the delayed response. I finally had some time to delve deeper into this issue and came to the same conclusion you have; there is indeed a sign error for the OSWEC Hydraulic PTO application case. I think this is due to the inclusion of **Rotary to Linear Adjustable Rod** and its sign convention (since that's the only difference between the RM3 and OSWEC Hydraulic PTO cases). 

My question for you is where did you add the minus sign for the PTO-torque in the OSWEC Hydraulic PTO case?

**Comment by kmruehl** [comment]
**Note to self...**

OSWEC Hydraulic PTO: Flap Velocity, and Sensed Actuator Torque. The torque and velocity have the same sign:
![oswec_vf](https://user-images.githubusercontent.com/4603379/42962284-7b683190-8b4e-11e8-958f-55f441036e2c.png)

RM3 Hydraulic PTO: Float velocity, relative velocity, and Sensed Actuator Force. The force and velocity have opposite signs:
![rm3_vf](https://user-images.githubusercontent.com/4603379/42962207-4ce8cfe6-8b4e-11e8-921f-33b3e81549d0.png)

Sensed Actuator Torque is the torque applied to the PTO and 'sensed' by the Simscape Revolute Joint:
![sensed_actuator_torque](https://user-images.githubusercontent.com/4603379/42962405-c7c1685e-8b4e-11e8-81ee-71cc6c50d9aa.PNG)

For the OSWEC, the Sensed Actuator Torque (`output.ptos.forceActuation(:,5)`) and PTO-Sim Torque (`output.ptosim.pistonNCF.force`) have the same sign but different magnitudes (due to linkage), and the Internal Mechanics Torque (`output.ptos.forceInternalMechanics(:,5)`) is zero (for pto.c and pto.k set to 0):
![oswec_f](https://user-images.githubusercontent.com/4603379/42962540-1666a398-8b4f-11e8-8cef-4023fce199b1.png)

For the RM3, the Sensed Actuator Force and PTO-Sim Force have the same sign and different magnitudes, and the Internal Mechanics Force is zero (for pto.c and pto.k set to 0):
![rm3_f](https://user-images.githubusercontent.com/4603379/42962638-4decd238-8b4f-11e8-9b14-ee7a8258563a.png)

Added a negative sign internal to the pto-sim subsystem, this corrects the sign error for the Sensed Actuator Torque `output.ptos.forceActuation(:,5)`, but not for the PTO-Sim Torque `output.ptosim.pistonNCF.force`:
![oswec_negsign](https://user-images.githubusercontent.com/4603379/42962867-dd055d82-8b4f-11e8-8618-88b9d48d807c.PNG)

Need to resolve issue to correct both the Sensed Actuator Torque (`output.ptos.forceActuation(:,5)`) and PTO-Sim Torque (`output.ptosim.pistonNCF.force`)

**Comment by liangsunwut** [comment]
Was this issue solved?

**Comment by jleonqu** [comment]
I spent some time working on this issue and I found an error in the Rotary to Linear Adjustable Rod block in PTO-Sim. The error is in the calculation of the torque of the hydraulic PTO. The hydraulic cylinder exerts a force on the flap, and this creates a torque on the pivot of the flap. I recalculated the torque by using a different equation based on the figure shown below:

![image](https://user-images.githubusercontent.com/72461917/101404815-6540e880-3894-11eb-8839-0a988adee816.png)

The torque in the point O can be calculated using this equation:

![image](https://user-images.githubusercontent.com/72461917/101413807-287bee00-38a2-11eb-837e-51115a143715.png)

The torque obtained using the new formulation and the flap velocity have opposite signs, as it was expected. This new formulation corrects the original results, when the torque and the flap velocity had the same sign.

![image](https://user-images.githubusercontent.com/72461917/101414501-6c232780-38a3-11eb-966c-5de8224cffc2.png)
